Initial material distribution

gloopydoop opened the pull request:
This one is pretty self explanatory.

But let me know your thoughts on moving the boundary_mesh utilities out of the `brinkman_source_term` and into `set_scalar_ic` in `neko`, and then we all have public access to that functionality.

I guess something also up for discussion is how multiple point zones is handled. Currently it's quite nice in the `brinkman_source_term` but I also saw that `point_zones` support boolean operations, so maybe this is the way forward.

https://neko.cfd/docs/develop/da/dd0/point-zones.html

Depends on #124

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into GPU_support/objective_function

gloopydoop committed:
linting?

gloopydoop committed:
lintinggg

gloopydoop committed:
I MUST have thwarted the linter by now!

gloopydoop committed:
I got you this timegit add topopt_design.f90!

gloopydoop committed:
move mapping

gloopydoop committed:
hopefully last one!

gloopydoop committed:
omg lastone....

gloopydoop committed:
stamping

gloopydoop committed:
how did I fail the linter on osmething so small...

gloopydoop committed:
update CMakeList

timfelle committed:
Relax the debug

timfelle committed:
Revert "Enhance Fortran build configuration and add LAPACK interface (#86)"

This reverts commit 1448f00ad375d8172e08f8ce66cd847b2a92bd8f, reversing
changes made to 4c836a86deccc3f3f1e4609a1492fd9c0cc8ac2c.

timfelle committed:
Discussions with Niclas led to some more finds in terms of issues.

(cherry picked from commit f56db72a933cd10ef764e80395a6d4b30af3b055)

timfelle committed:
Work on array temporaries

AbbasMoosaV committed:
Update CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Update CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Force run the CI for debug purpose

timfelle committed:
Don't cache system for pfunit

timfelle committed:
same for gslib

timfelle committed:
Reactivate cache and improve parallelism

timfelle committed:
Move checks about a bit, but it should be good now

timfelle committed:
Commenting update to improve readability of workflow

AbbasMoosaV committed:
Updates based on the PR comments.

timfelle committed:
forgot to remove check from main workflow

timfelle committed:
test without cache on neko

timfelle committed:
Once again

timfelle committed:
Reactivate the cache

timfelle committed:
un-cache json

timfelle committed:
skip caching for system

AbbasMoosaV committed:
Updates based on the PR comments 2.

timfelle committed:
reactivate cache

timfelle committed:
Apt cache is broken

timfelle committed:
Revert Fortran build configuration changes and improve debugging options (#89)

Revert previous enhancements to the Fortran build configuration while
relaxing debug options and addressing identified issues with array
temporaries.

timfelle committed:
MMA loop tweeks (#88)

Very small PR, but pretty important.

The design should be mapped to the brinkman term in between the MMA
update and running the simulation.

Also, stamping the zeroth iteration is nice for visualization.

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into GPU_support/objective_function

gloopydoop committed:
linting

gloopydoop committed:
Update sources/designs/topopt_design.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Update sources/mapping_functions/PDE_filter_mapping.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Update sources/mapping_functions/PDE_filter_mapping.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
PR comments

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

gloopydoop committed:
trailing spaces

timfelle committed:
Massive improvement to the cmake structure.

- Neko-TOP static library is built by default
- The `Examples` target now builds the examples.
- The setup script is updated to account for this.

timfelle committed:
Improve CMake structure and build process for examples (#90)

- Neko-TOP static library is built by default
- The `Examples` target now builds the examples.
- The setup script is updated to account for this.

Time to run `./setup.sh` (With all external dependencies ready)

- Before: 41 sec
- After: 26 sec

gloopydoop committed:
first draft

gloopydoop committed:
little adjustment

gloopydoop committed:
Update sources/CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Declare in same order as usage in the files

timfelle committed:
Move MMA settings into the solver section

timfelle committed:
Small clean up of optimizer

timfelle committed:
Check against the result folder instead of a log

timfelle committed:
Add a factory for generic optimizer

timfelle committed:
Fix uninitialized variables

timfelle committed:
Update driver with the parameter read

timfelle committed:
Simplify optimizer (Invalidated previous results)

timfelle committed:
Updates are functional

timfelle committed:
Slight cleanup

timfelle committed:
Add base initializer

timfelle committed:
Updates to our interfaces

timfelle committed:
Align the interfaces with style from neko

timfelle committed:
Improve the driver structure.

timfelle committed:
Use the simulation reset function

timfelle committed:
Update parameter order

timfelle committed:
Add a file reader function for json

gloopydoop committed:
Update sources/neko_ext/math/bcknd/device/cuda/math_ext_kernel.h

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Update sources/problems/objectives/lube_term_objective.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
PR suggestions

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

AbbasMoosaV committed:
Restructure mma_t removing the mma_factory and utilizing submodules for the backend. Cuda and CPU backends checked for example/mma. Further cleanups needed.

gloopydoop committed:
PR comments

gloopydoop committed:
Update sources/problems/objectives/lube_term_objective.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Apply suggestions from code review

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
import c_ptr, c_int

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

gloopydoop committed:
fromatting

gloopydoop committed:
slight updates

gloopydoop committed:
minor changes

gloopydoop committed:
csv_file_t

gloopydoop committed:
PR comments

gloopydoop committed:
lint

gloopydoop committed:
PR comments

gloopydoop committed:
.txt -> .csv

gloopydoop committed:
formatting

gloopydoop committed:
back to the zeroth iteration

gloopydoop committed:
Update sources/neko_ext/math/bcknd/device/cuda/math_ext_kernel.h

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
initial condition

timfelle committed:
Improved logging (#91)

I guess we discussed this a bit this morning, but this is what I had in
mind.
I'm putting this in draft mode because I want comments before I move any
further with this:

- Do we think we'll want this in other optimization? ie, would a
`logger_t` be useful?
- Are we all happy with enforcing a log name to all
constraints/objectives so we can write them out?

But perhaps more importantly, what I've written currently is pretty
susceptible to truncation of that long string (buffer? is this the
correct word? that was the intention of the variable name `buff`)

### Example output

```
n =     86400, m =         1, tolerance =  1.00000000000000002E-03
iter, Total objective function, Dissipation, out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
  0,     5.3696209634E+01,     4.0441625841E+01,     1.3254583793E+01,    -3.4113716238E-01,     1.7976931349+308,     1.7976931349+308,     1.0000000000E+00
  1,     5.3631210218E+01,     4.0441625841E+01,     1.3189584378E+01,    -3.4067827356E-01,     3.4063388244E-01,     1.8251180684E+00,     1.0000000000E+00
  2,     5.3266158113E+01,     4.0127814834E+01,     1.3138343279E+01,    -3.4021952588E-01,     3.4017508569E-01,     5.0791113981E+00,     1.0000000000E+00
  3,     5.2946050791E+01,     3.9894493099E+01,     1.3051557691E+01,    -3.3973026153E-01,     3.3968611728E-01,     4.1537413064E+00,     1.0000000000E+00
  4,     5.2488335978E+01,     3.9561169448E+01,     1.2927166530E+01,    -3.3922171690E-01,     3.3917795103E-01,     5.6123647507E+00,     1.0000000000E+00
  5,     5.1825567047E+01,     3.9082081599E+01,     1.2743485448E+01,    -3.3889046816E-01,     3.3884813432E-01,     7.9908112190E+00,     1.0000000000E+00
  6,     5.1000305086E+01,     3.8491289553E+01,     1.2509015532E+01,    -3.3852679847E-01,     3.3848677945E-01,     9.8497959414E+00,     1.0000000000E+00
  7,     5.0354702281E+01,     3.8025030555E+01,     1.2329671726E+01,    -3.3866646158E-01,     3.3862891930E-01,     7.8845237703E+00,     1.0000000000E+00
  8,     4.9785292383E+01,     3.7615237248E+01,     1.2170055135E+01,    -3.3902719413E-01,     3.3899308096E-01,     7.5623243319E+00,     1.0000000000E+00

```

### Example output using `csv_file_type`
```
iter, Total objective function, Dissipation, Out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
0.0000000000000000,53.696209634065141,40.441625840817053,13.254583793248088,-0.34113716238378983,0.0000000000000000,0.0000000000000000,1.0000000000000000
1.0000000000000000,53.631210218451116,40.441625840817053,13.189584377634059,-0.34067827356064134,0.34063388243730330,1.8251180684119177,1.0000000000000000
2.0000000000000000,53.266158113095081,40.127814833617478,13.138343279477603,-0.34021952587850796,0.34017508568651300,5.0791113980976883,1.0000000000000000
3.0000000000000000,52.946050790610592,39.894493099386999,13.051557691223593,-0.33973026152689706,0.33968611727815884,4.1537413063900503,1.0000000000000000
4.0000000000000000,52.488335977948438,39.561169447724346,12.927166530224092,-0.33922171690237829,0.33917795102621462,5.6123647507375267,1.0000000000000000
5.0000000000000000,51.825567047220396,39.082081599088639,12.743485448131759,-0.33889046815908813,0.33884813432446081,7.9908112190203324,1.0000000000000000
6.0000000000000000,51.000305085635944,38.491289553207920,12.509015532428020,-0.33852679846645700,0.33848677945129146,9.8497959414055956,1.0000000000000000
7.0000000000000000,50.354702281281916,38.025030555367294,12.329671725914622,-0.33866646158438390,0.33862891930457578,7.8845237703490136,1.0000000000000000
8.0000000000000000,49.785292383215989,37.615237248149093,12.170055135066898,-0.33902719412641674,0.33899308096021136,7.5623243318792319,1.0000000000000000
```

gloopydoop committed:
merge develop

timfelle committed:
Merge commit '3f11b506ff4b802167a10aafb58841c947ea526f' into driver/json

timfelle committed:
Merge commit '7a0b303bcd03c05e12edd54a976826a89824f2b2' into driver/json

AbbasMoosaV committed:
Cleanups and updates based on PR comments.

gloopydoop committed:
Apply suggestions from code review

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
PR comments

timfelle committed:
Apply suggestions from code review

AbbasMoosaV committed:
Update the structure of mma.cu.

timfelle committed:
Fix the possibility to specify the number of processors to run with

timfelle committed:
Allow specification of processor count for execution (#92)

Introduce the ability to specify the number of processors to use via the
`NPROCS` environment variable, enhancing flexibility in resource
allocation during script execution.

AbbasMoosaV committed:
Update the structure of mma_kernel.h

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into GPU_support/objective_function

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

gloopydoop committed:
mma

timfelle committed:
Enhance JSON handling and driver structure improvements (#72)

Introduce json based initialization for all main types.

Additionally I have split the simulation out of the problem in order to
simplify dependencies.

Depends on #76

timfelle committed:
Merge branch 'develop' into GPU_support/objective_function

timfelle committed:
Fix formatting

timfelle committed:
Up the linting score a bit in general

timfelle committed:
Gpu support (#79)

Originally this was 5 branches, and I wanted to submit 5 individual PRs
so this would be a bit cleaner.
But essentially this is bringing the Rugby ball case up to date with GPU
support.
The three branches were:
`GPU_support/masks`
`GPU_support/volume_constraint`
`GPU_support/objective_function`
`GPU_support/design` (which also contained the RAMP mappings etc)
`GPU_support/MMA_on_CPU` This one won't be needed, but it essentially
mem_copied everything from device to host, did MMA on the CPU, then
copied back.

But I was able to run the rugby ball on GPU so I was quite happy!
However, the optimization history wasn't EXACTLY the same as CPUs, so
I'm keeping this on draft until I can test more and resolve this.

**EDIT**

Ok I'm more and more confident the problem is with MMA on CPU, so I
think we can take this off draft mode. Also it looks like PR#85 needs a
lot of revisions and I want to get this PR through so we can all start
testing things on GPU. I apologies if there is bug is in here unrelated
to the hacky memcopies I performed.

 **note:**
I made a rather fundamental change to PDE filter, ie, solving for
$$\Delta \tilde{\rho}$$ not $$\tilde{\rho}$$, so already we won't match
with develop exactly, due to the residual prescribed by the PDE filter.

### Convergence history on develop
```
  0,  5.36962096340651414E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962096340651414E+01, -3.40678273560641343E-01,  3.40633882437303304E-01,  1.82511436798967752E+00,  1.00000000000000000E+00
  2,  5.34909202977577962E+01, -3.40219526392907701E-01,  3.40175086197066656E-01,  2.93977673685759111E+00,  1.00000000000000000E+00
  3,  5.32662799775875442E+01, -3.39695799113361951E-01,  3.39651438883154300E-01,  3.19543034281953853E+00,  1.00000000000000000E+00
  4,  5.29593114254205517E+01, -3.39109389875194511E-01,  3.39065248562431587E-01,  3.98656352093350419E+00,  1.00000000000000000E+00
  5,  5.25307111679251335E+01, -3.38585236396372891E-01,  3.38541901793160127E-01,  5.22808835345366951E+00,  1.00000000000000000E+00
  6,  5.19234539710996330E+01, -3.38143391208646948E-01,  3.38102175627622070E-01,  7.03658279719623980E+00,  1.00000000000000000E+00
  7,  5.11597748945414921E+01, -3.37928412343896978E-01,  3.37888261694012759E-01,  8.82061398036337785E+00,  1.00000000000000000E+00
  8,  5.05602700893721249E+01, -3.38296226436742675E-01,  3.38259998372253234E-01,  6.98241437563093026E+00,  1.00000000000000000E+00
  9,  5.00465163087089451E+01, -3.38384508012592067E-01,  3.38349557059135442E-01,  6.54315441819684906E+00,  1.00000000000000000E+00
 10,  4.95253467948585779E+01, -3.38395758796641910E-01,  3.38362432950295389E-01,  6.75821664949387291E+00,  1.00000000000000000E+00
```

### Convergence history on CPU (with the new changes)
```
  0,  5.36962098917103532E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917103532E+01, -3.40678272361158607E-01,  3.40633881225604174E-01,  1.82511372436923880E+00,  1.00000000000000000E+00
  2,  5.34909201511379564E+01, -3.40219524004039708E-01,  3.40175083796164401E-01,  2.93978525062977747E+00,  1.00000000000000000E+00
  3,  5.32662797551102756E+01, -3.39695795330679007E-01,  3.39651435093588583E-01,  3.19540660892341366E+00,  1.00000000000000000E+00
  4,  5.29593111639978673E+01, -3.39109383832420674E-01,  3.39065242501398190E-01,  3.98658874364438764E+00,  1.00000000000000000E+00
  5,  5.25307107108303555E+01, -3.38585227652321519E-01,  3.38541893041661768E-01,  5.22806768673648037E+00,  1.00000000000000000E+00
  6,  5.19234524546626943E+01, -3.38143379103141128E-01,  3.38102163503164688E-01,  7.03658086336280775E+00,  1.00000000000000000E+00
  7,  5.11597740019539202E+01, -3.37928397133592295E-01,  3.37888246463098230E-01,  8.82063926519529318E+00,  1.00000000000000000E+00
  8,  5.05602689953278457E+01, -3.38296209449853336E-01,  3.38259981376574370E-01,  6.98237874556235294E+00,  1.00000000000000000E+00
  9,  5.00465151924080729E+01, -3.38384488855007104E-01,  3.38349537884281071E-01,  6.54318127887768597E+00,  1.00000000000000000E+00
 10,  4.95253453940975277E+01, -3.38395744127610365E-01,  3.38362418271727805E-01,  6.75818420810124021E+00,  1.00000000000000000E+00
```

### Convergence history on GPU (with the new changes)
```
  0,  5.36962098917055499E+01, -3.41137162383794046E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917055499E+01, -3.40678272944622995E-01,  6.26706799280533833E+00,  6.43310832024656179E+02,  1.00000000000000000E+00
  2,  5.34910118155738559E+01, -3.40219525367835784E-01,  6.26705020196007023E+00,  6.43309822428294638E+02,  1.00000000000000000E+00
  3,  5.32664717580504288E+01, -3.39691202134478287E-01,  6.23507603889123274E+00,  6.41924745328890140E+02,  1.00000000000000000E+00
  4,  5.30975501242981096E+01, -3.39151984540635609E-01,  6.19951479473746137E+00,  6.40112443825842547E+02,  1.00000000000000000E+00
  5,  5.25350413808810828E+01, -3.38644965953474764E-01,  6.16274815428317435E+00,  6.38611032141696114E+02,  1.00000000000000000E+00
  6,  5.19248549714847698E+01, -3.38207041002288433E-01,  6.08546842737498661E+00,  6.34741181740812067E+02,  1.00000000000000000E+00
  7,  5.11622343791975069E+01, -3.37997668480812863E-01,  5.98907463234943904E+00,  6.30183729381854391E+02,  1.00000000000000000E+00
  8,  5.05718316461075759E+01, -3.38299594367291456E-01,  5.86614026930082400E+00,  6.24306904330888301E+02,  1.00000000000000000E+00
  9,  5.00919821259184630E+01, -3.38447606977853088E-01,  5.77873574707956283E+00,  6.19710815500791000E+02,  1.00000000000000000E+00
 10,  4.96736731441919943E+01, -3.38192325940090688E-01,  5.69855410966845621E+00,  6.15708752891213408E+02,  1.00000000000000000E+00
```

timfelle committed:
Add missing flag to cmake.

Lost when Testing config was added

timfelle committed:
Build the examples from the runner script

timfelle committed:
Fix a few things in the verification run script

timfelle committed:
Move the source term to design.

timfelle committed:
Print MPI safe

timfelle committed:
Move mapping to parent design

timfelle committed:
Move computations to problem

timfelle committed:
Move design output to design.

timfelle committed:
Move the last pieces out of the steady state problem, hence it is removed.

timfelle committed:
revert change

timfelle committed:
Update sources/optimizer/mma_optimizer.f90

gloopydoop committed:
missing import

gloopydoop committed:
half way there

gloopydoop committed:
Merge branch 'bugfix/memcopy' into fix/rugby_ball_case

timfelle committed:
missing import in `mma_optimizer` (#96)

missing import

gloopydoop committed:
and the driver too

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into fix/rugby_ball_case

AbbasMoosaV committed:
Merge develop into feature/gpu_support_MMA

AbbasMoosaV committed:
Add error for unknown backend in mma_init_attribute based on PR comments.

timfelle committed:
Fix verification run script and build examples from runner script (#94)

Streamline the execution process by building example executables from
the runner script.

Update the verification example. New filenames were not used properly.
Also the example will now check against the result folder, rather than a
manually added reference.

Also added an MPI guard on a few print statements

timfelle committed:
Bring rugby ball up to date (#97)

Here I've just updated the case file and driver of `rugby_ball` to
reflect the to updates in `rugby_verification`

**NOTE**

This will fail on device at the 2nd iteration due to the KKT check at
line 413 of `mma.f90`. Substituting
```fortran
    if (NEKO_BCKND_DEVICE .eq. 0) then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

with
``` fortran
    if (this%backend == 'cpu') then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

Is how PR#79 was run

AbbasMoosaV committed:
Updates based on suggested improvments (Linting).

AbbasMoosaV committed:
Improving the format.

AbbasMoosaV committed:
Resolve compiler warnings.

AbbasMoosaV committed:
Merge develop and resolve conflicts.

AbbasMoosaV committed:
Correct the (only :) typo in all files in the source folder.

AbbasMoosaV committed:
Linting updates.

AbbasMoosaV committed:
More linting updates.

AbbasMoosaV committed:
Linting!

timfelle committed:
Formatting

AbbasMoosaV committed:
Merge mma%free%device and mma%free%cpu to mma%free.

gloopydoop committed:
time to test the linting

gloopydoop committed:
linting

gloopydoop committed:
linting

gloopydoop committed:
forgot the cmake

AbbasMoosaV committed:
Updates based on PR comments 1.

timfelle committed:
Linter updates

timfelle committed:
fix bug

gloopydoop committed:
up the linting

gloopydoop committed:
maybe the docs?

timfelle committed:
Up the score

timfelle committed:
upgrade flinter for the CI

timfelle committed:
fix requirements

timfelle committed:
revert changes to improve compatibility with other PR's. Simplifying the merge queue

gloopydoop committed:
suggestion for field writing

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into refactor/design-problem

gloopydoop committed:
merge develop and name change

timfelle committed:
Refactor/design-problem (#93)

This PR build on the structure.

The design, problem and simulation is untangled. The steady state
problem are no longer needed.

gloopydoop committed:
u_adj

gloopydoop committed:
unused import

gloopydoop committed:
Merge branch 'docs/adjoint' of github.com:ExtremeFLOW/neko-top into docs/adjoint

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into docs/adjoint

AbbasMoosaV committed:
Updates based on PR comments 2.

gloopydoop committed:
links

gloopydoop committed:
Update sources/CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Updates based on PR comments 3.

AbbasMoosaV committed:
Updates based on the PR comments 4.

AbbasMoosaV committed:
Updates based on PR comments 5.

gloopydoop committed:
forgot to add adjoint.md

gloopydoop committed:
Merge branch 'docs/adjoint' of github.com:ExtremeFLOW/neko-top into docs/adjoint

gloopydoop committed:
big equations good, inline equations bad

gloopydoop committed:
few!

timfelle committed:
Documentation for adjoint fluid + name changes (#99)

The PR is moving in the direction of getting the adjoint passive scalar
off the ground. It consists of:
1) A bit of documentation for the adjoint velocity equations (I'll fill
in the adjoint passive scalar in a future PR). It isn't linked to the
rest of the docs, I'll wait for PR#95.
2) A lot of name changes such that `adjoint` becomes `adjoint_fluid`
where applicable, opening the way for us to have
- `fluid` and `adjoint_fluid`
- `scalar` and `adjoint_scalar`

timfelle committed:
Revert "Documentation for adjoint fluid + name changes (#99)"

This reverts commit e86bc8fcf9e83556e64f4769573ae9840bb8cadf, reversing
changes made to 4a78019056cd8d11e5e3a9b0030b3ee287e08f07.

timfelle committed:
formatting

timfelle committed:
Revert documentation changes and update adjoint references (#100)

Revert previous documentation changes related to adjoint fluid and
update references to use a unified adjoint scheme. Adjustments include
modifications to various modules and removal of outdated documentation.

gloopydoop committed:
wooooowwwwww that was close

gloopydoop committed:
these guys were in comments, but still to be safe

AbbasMoosaV committed:
Merge in develop and resolve conflicts.

AbbasMoosaV committed:
Updates based on PR comments 6.

AbbasMoosaV committed:
Little formatting.

AbbasMoosaV committed:
Format!

AbbasMoosaV committed:
Formatting!

AbbasMoosaV committed:
Formatting!

gloopydoop committed:
Revert "Revert documentation changes and update adjoint references"

gloopydoop committed:
Merge commit 'cb271fb40e194239329387628bee1a79d98daed9' into revert-100-revert/adjoint_update

gloopydoop committed:
Merge commit '875569046c8b2edf384e2660dd547adcb06a2b2c' into revert-100-revert/adjoint_update

gloopydoop committed:
linting

gloopydoop committed:
linting

timfelle committed:
update cleanup function

timfelle committed:
Apply suggestion

timfelle committed:
this is what chat-GPT says to do :sweat_smile:  (#102)

what a kerfuffle...

according to chatGPT:

If you reverted the original PR using another PR, GitHub provides a
"Revert" button:

Go to the PR that reverted your original changes.
Click on the "Revert" button (if available). This will create a new PR
that restores the original changes.
Then, apply your fix on top of that before merging.

I hope that's what has been done in this PR, but please double check

timfelle committed:
Merge remote-tracking branch 'origin/develop' into feature/gpu_support_MMA

timfelle committed:
Lint

timfelle committed:
lint

timfelle committed:
Feature/gpu support mma (#85)

This updates the mma_optimizer structure, introduces mma_factory for CPU
and device backend types and generally, adds CUDA support for mma
solver.

timfelle committed:
Make sure to not use built-in names for variables

timfelle committed:
Exit with an error if examples cannot be compiled

timfelle committed:
Avoid built-in names and handle compilation errors (#104)

Ensure variable names do not conflict with built-in names and exit with
an error message if examples fail to compile.

gloopydoop committed:
small change

gloopydoop committed:
strange line

gloopydoop committed:
still that strange insertion

timfelle committed:
really small change (#108)

I think the changes are obvious.
But I'd never seen that `"output_control": "never"` before. That's
really nice for us :+1:

gloopydoop committed:
Update for `neko` event (#111)

Update based on `neko` PR #1736

gloopydoop committed:
remove premultiply adjoint forcing with mass matrix (#112)

The pre-multiplication is handled in the source term handler, we
shouldn't have it here.

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Solve the mpi reduce issue in subsolve in mma_device. (#113)

Small fix in  mma_device.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Refactor the design_t (#103)

Major update to the design type.
This PR will improve the reuseability of the design type by hiding
implementation details behind a generic API.

gloopydoop committed:
whole lot of work

gloopydoop committed:
forgotten files

timfelle committed:
Sync formatting to neko standards (#125)

Update formatting commands to align with neko's specifications.

Continuation lines starting with a `&` is not indented to 0 compared to
their current scope.

This is commonly used to break long strings.

gloopydoop committed:
Fix fluid ICs (#126)

This is in response to my `neko` PR passing the subdict to ICs.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
something went wrong

timfelle commented:
@gloopydoop Did you mean to target the mapping PR?

gloopydoop commented:
> @gloopydoop Did you mean to target the mapping PR?

I did because I'd already pulled that branch into this branch and the files changed looked cleaner this way :sweat_smile: . Both of them are still darft, I have some final testing and tinkering to do. I most made this PR to see if you agree with moving a lot of the `brinkman_source_term` stuff over to `scalar_ic` on the `neko` side such that those routines are a bit more public. The alternative is we write a brand new set of routines for setting the initial material distribution but that feels like a lot of double of code.

On a side note, at some point we need to write a `neko-top` equivalent of `user_inf.f90`, I think user defined initial material distributions would be a valuable addition.

gloopydoop committed:
one down

timfelle committed:
Align with new KSP in neko (#129)

Update the usage of the Krylov solver to align with the new KSP
interface, replacing calls to `krylov_solver_destroy` with the
appropriate `free` method for memory management.

timfelle committed:
Cleanup and fix examples in the repository (#128)

Remove outdated and broken example files while fixing existing examples
to ensure proper functionality. This cleanup enhances the overall
quality and maintainability of the codebase.

gloopydoop committed:
fix mistake

gloopydoop committed:
Merge branch 'feature/mapping' of github.com:ExtremeFLOW/neko-top into feature/mapping

gloopydoop committed:
Merge branch 'develop' into feature/mapping

gloopydoop committed:
linting

gloopydoop committed:
Merge branch 'feature/mapping' of github.com:ExtremeFLOW/neko-top into feature/mapping

timfelle committed:
Documentation improvements and example updates (#95)

Enhance documentation clarity and structure by updating references,
fixing formatting issues, and removing redundant content. Add missing
examples and address warnings in the documentation.

gloopydoop committed:
segregate initial distribution

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into feature/initial_matieral_distro

gloopydoop committed:
convention for design living in u

timfelle commented on sources/designs/design_brinkman.f90:
```diff
@@ -214,16 +229,33 @@ subroutine brinkman_design_init_from_json_sim(this, parameters, simulation)
+      call json_extract_object(parameters, &
+           'optimization.design.initial_distribution', json_subdict)
+
+      if (trim(string_val) .ne. 'user') then
+         call set_optimization_ic(this%design_indicator, coef, gs, string_val, &
+            json_subdict)
+      else
+         call neko_error("user defined initial material distrubtions should &
+         & be enforced through the driver.")
+      end if
```
We will not spent time implementing a user system similar to the one in neko.
There is no benefit from doing so.
Also why are you passing the string_value in? The exact information is also
contained in the json_subdict.
```suggestion
      call set_optimization_ic(this%design_indicator, coef, gs, string_val, &
           json_subdict)
```

gloopydoop replied:
take another look, I figured we should also start with zero if nothing is prescribed

timfelle committed:
Align with new checkpoint structure in Neko (#133)

Update the code to align with the new checkpoint structure, ensuring
that time-related parameters are accessed correctly throughout the
adjoint and reset subroutines.

gloopydoop committed:
formatting

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into feature/initial_matieral_distro

gloopydoop committed:
clean optimization_ic

gloopydoop committed:
clean typos

gloopydoop committed:
formatting

timfelle committed:
Ensure cross compilation support and clean up build flags (#122)

Introduce a new file for compiler flags to streamline the build process
and ensure compatibility with cross compilation. This change simplifies
the management of compiler options for different build types.

timfelle committed:
Fix aliasing module references (#123)

Update module references from `simulation` to `simulation_m` to resolve
aliasing issues. Adjust initialization and cleanup processes
accordingly.

Fixes: #116

gloopydoop committed:
Definition of steady residual + logging (#131)

I'm trying to dissect some of the changes that were originally in #110 .

Here I've changed the definition of the steady residual and added a
little csv writer so it can be monitored

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,373 @@
+
+    call json_get(params, "type", type)
+
+    if (trim(type) .eq. 'uniform') then
```
Please use `select case` for this kind of switching, it is a lot cleaner.

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,373 @@
+
+    else if (trim(type) .eq. 'field') then
+
+       call json_get(params, 'file_name', read_str)
+       fname = trim(read_str)
```
How about trimming inside the function instead, then you also safeguard other
users. and you can read directly into thevariable

gloopydoop replied:
I'm not sure I understand completely, and it was Victor who implemented `set_scalar_IC` so maybe I can talk to him when I get back to Stockholm, but those two names were declared rather deliberatly as:
```fortran 
character(len=NEKO_FNAME_LEN) :: fname, mesh_fname
``` 
so he must have had his reasons

timfelle replied:
My suggestion is, remove trim here

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,373 @@
+       call json_get_or_default(params, 'interpolate', interpolate, &
+            .false.)
+       call json_get_or_default(params, 'tolerance', tol, 0.000001_rp)
+       call json_get_or_default(params, 'mesh_file_name', read_str, &
+            "none")
```
Does this not fit in one line?

gloopydoop committed:
PR comment select case

gloopydoop committed:
one line

gloopydoop committed:
Mapping cascade (#124)

This PR has a couple of nice features, but is still a bit of a work in
progress (maybe)

- The main contribution is the notion of a mapping cascade and a lot of
JSON support for the pre-existing mappings. Now we prescribe a list of
mappings which are performed sequentially, and it also takes care of the
chain rule moving backwards.
- There are some cosmetic changes to variable names, particularly
`sens_in` and `sens_out` :sweat_smile:
- Obviously much of the mapping was inspired by the source term handler,
so in principal this could be extended to allow user defined mappings in
some way, but this is to be discussed.

This PR is still in draft mode because I think we should really discuss
who owns the mapping, and how we should handle mapping to multiple
coeffients, and how we should hand constraints defined on intermediate
steps (ie, the potential for the volume constraint to defined on the
filtered design variable)

There are still a few cosmetics changes that need to be done so I'm
leaving this on draft mode, but take a look if you can and tell me what
you think

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
believe develop

gloopydoop committed:
merge develop

gloopydoop committed:
remove unused files

gloopydoop commented on sources/neko_ext/neko_ext.f90:
```diff
@@ -24,6 +24,14 @@ module neko_ext
+  use field, only: field_t
+  use utils, only: neko_error
+  use json_module, only : json_file
   implicit none
```
I don't know how #124 got through without this? I didn't do the best job of splitting things apart :sweat_smile:

gloopydoop committed:
oh that's why

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,371 @@
+    end if
+
+    ! Change from "field0.f000*" to "field0.fld" for the fld reader
+    call filename_chsuffix(file_name, file_name, 'fld')
```
Trim here instead

timfelle commented on examples/rugby_ball/rugby_ball.case:
```diff
@@ -86,7 +86,14 @@
                 ],
-            }
+            },
+            {
+                "name": "initial_ball",
+                "geometry": "cylinder",
+                "start": [0.5, 0.5, -0.5],
+                "end": [0.5, 0.5, 1.5],
+                "radius": 0.25
+    },
```
```suggestion
            {
                "name": "initial_ball",
                "geometry": "cylinder",
                "start": [0.5, 0.5, -0.5],
                "end": [0.5, 0.5, 1.5],
                "radius": 0.25
            },
```

timfelle commented on examples/rugby_verification/rugby_ball.case:
```diff
@@ -81,7 +81,14 @@
                 ],
-            }
+            },
+            {
+                "name": "initial_ball",
+                "geometry": "cylinder",
+                "start": [0.5, 0.5, -0.5],
+                "end": [0.5, 0.5, 1.5],
+                "radius": 0.25
+    },
```
```suggestion
            {
                "name": "initial_ball",
                "geometry": "cylinder",
                "start": [0.5, 0.5, -0.5],
                "end": [0.5, 0.5, 1.5],
                "radius": 0.25
            },
```

timfelle commented on sources/designs/design_brinkman.f90:
```diff
@@ -227,16 +231,29 @@ subroutine brinkman_design_init_from_json_sim(this, parameters, simulation)
       call this%mapping%init_base(coef)
       call this%mapping%add(parameters, 'optimization.design.mapping')
+
+      if (parameters%valid_path(&
+           'optimization.design.initial_distribution')) then
```
Lets make this mandatory and allow the user to specify a "uniform" distribution
as well

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,371 @@
+    class(point_zone_t), pointer :: zone
+    integer :: size
+
+    call neko_log%message("Type       : point_zone")
```
Why all the spaces after type? it does not even match with the number of
characters in the "Base value" line just under it.

timfelle commented on sources/designs/set_optimization_ic.f90:
```diff
@@ -0,0 +1,371 @@
+       ! Do the space-to-space interpolation
+       call space_interp%map_host(fld%x, fld_data%t%x, fld_data%nelv, fld%Xh)
+
+       call space_interp%free
```
Please don't blindly copy Neko. 
Add parenthesis for function calls, even when no parameters are given.

gloopydoop committed:
Optimization domain in JSON (#139)

Wow I can't believe this one was still hardcoded :sweat_smile: 
I've put the "help wanted" flag on because I don't know how this
translates into an optimization problem without a simulation. I guess
the word "domain" for topology optimization means a spatial domain, so
to me I think it should probably be rooted in
`optimization.design.domain`

But it's worth discussing.

Also aiming for many small PRs moving forward :sweat_smile:

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Improve documentation and fix formatting issues (#138)

Enhance the documentation configuration and deployment, correct markdown
formatting, and complete doxygen comments across multiple files.
Addressed bugs in doxygen and made spelling updates for clarity.

timfelle committed:
Fix segmentation fault and improve logging (#134)

Address segmentation faults by ensuring proper memory management and
initialization. Enhance logging capabilities for better debugging and
add tests to verify functionality. Align code with new checkpoint
structures and clean up unnecessary outputs.

---------

Co-authored-by: Harrison Nobis <58177854+gloopydoop@users.noreply.github.com>

gloopydoop committed:
brackets

gloopydoop committed:
Update examples/rugby_ball/rugby_ball.case

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Update examples/rugby_verification/rugby_ball.case

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Merge branch 'develop' into dev/cleanup

timfelle committed:
Merge branch 'develop' into feature/initial_matieral_distro

timfelle committed:
tiny fix

timfelle committed:
Merge commit 'f6fe84296215950df3d343027bdb9168e8c5b4a6' into feature/initial_matieral_distro

timfelle committed:
fix doc

timfelle committed:
Fix prints
